### PR TITLE
DAH-2858 feat: sort fcfs sales and render below open listings

### PR DESCRIPTION
**Pull request by cade-exygy**
## Description

- When displaying listings on the sales directory, check if a listing is a FCFS Sales listing, and then sort that listing by descending application start date
- The FCFS Sales listings should be behind the FCFS feature flag.

NOTE: Reviewer should test both tickets when reviewing this PR.

## Jira ticket

https://sfgovdt.jira.com/browse/DAH-2858
https://sfgovdt.jira.com/browse/DAH-2910

### Version Control

- [x] branch name begins with `angular` if it contains updates to Angular code
- [x] branch name contains the Jira ticket number
- [x] PR name follows `type: TICKET-NUMBER Description` format, e.g. `feat: DAH-123 New Feature`

### Code quality

- [x] [the set of changes is small](https://google.github.io/eng-practices/review/developer/small-cls.html#what-is-small)
- [x] all automated code checks pass (linting, tests, coverage, etc.)
- [x] code irrelevant to the ticket is not modified e.g. changing indentation due to automated formatting
- [x] if the code changes the UI, it matches the UI design exactly

### Review instructions

- [x] instructions specify which environment(s) it applies to
- [x] instructions work for PA testers
- [x] instructions have already been performed at least once

### Request review

- [x] PR has `needs review` label
- [x] Use `Housing Eng` group to automatically assign reviewers, and/or assign specific engineers
- [x] If time sensitive, notify engineers in Slack

## Review Instructions
- Go to Sales directory
- Ensure FCFS Sales listings show up under the open listings (ORA Should be the first one)
- Ensure that FCFS Sales listings are in the correct order (newest Start Date first)
- If you want, try updating the Start Date of a FCFS Sales listing in Salesforce and ensure the changes are reflected in the displayed sort order

### CODE REVIEW

File: app/javascript/modules/listings/DirectoryHelpers.tsx
Comment (cade-exygy) [inline app/javascript/modules/listings/DirectoryHelpers.tsx:251]: **question (non-blocking):**  So currently this is the same view as openListings. I am assuming we will want to make some visual changes for fcfs listings?

File: app/javascript/modules/listings/DirectoryHelpers.tsx
Comment (cade-exygy) [inline app/javascript/modules/listings/DirectoryHelpers.tsx:397]: **note:** Here is the order that results from this sort:
Screenshot 2024-11-07 at 11 34 19 AM

File: app/javascript/modules/listings/DirectoryHelpers.tsx
Comment (tallulahkay) [reply app/javascript/modules/listings/DirectoryHelpers.tsx:251]: no idts - it is supposed to be in the same view, so that the listings are all the lottery listings and then all the fcfs listings in one view

Comment (alulabeshue-sfgov) [reply app/javascript/modules/listings/DirectoryHelpers.tsx:251]: **note:** The only difference between fcfs sales and lottery listings will be the image card status but that is handled in `getListingCards`. Should we reuse `openListingsView`?

File: app/javascript/modules/listings/DirectoryHelpers.tsx
Comment (alulabeshue-sfgov) [reply app/javascript/modules/listings/DirectoryHelpers.tsx:397]: **issue:** I am seeing closed fcfs listings showing up on the directory page which the ACs on the ticket mentions they should not. Can we also add some tests for this logic?

File: app/javascript/modules/listings/DirectoryHelpers.tsx
Comment (cade-exygy) [reply app/javascript/modules/listings/DirectoryHelpers.tsx:251]: I'm cool with that, if that is clear as a reader.

File: app/javascript/modules/listings/DirectoryHelpers.tsx
Comment (cade-exygy) [reply app/javascript/modules/listings/DirectoryHelpers.tsx:397]: Yep we talked through that yesterday at refinement, which resulted in that clarification

File: app/javascript/modules/listings/DirectoryHelpers.tsx
Comment (cade-exygy) [reply app/javascript/modules/listings/DirectoryHelpers.tsx:251]: My main reason for creating a new function was readability, since fcfsListings are logically distinct from open in how we sort them.

Comment (alulabeshue-sfgov) [reply app/javascript/modules/listings/DirectoryHelpers.tsx:251]: Makes sense! I can go either way, Im ok with keeping the function for readability.

Comment (tallulahkay) [reply app/javascript/modules/listings/DirectoryHelpers.tsx:251]: Maybe call it openListingsWithFcfs though? bc it's not actually just fcfs listings

Comment (cade-exygy) [reply app/javascript/modules/listings/DirectoryHelpers.tsx:251]: the listings.fcfsSales should only contain fcfsListings. They display below listings.open. Here is how they are separated out before render:
```
// Sort listings in four buckets based on their status and filters
export const sortListings = (
  listings: RailsListing[],
  filters: EligibilityFilters,
  setMatch: React.Dispatch<React.SetStateAction<boolean>>
) => {
  const open = []
  const upcoming = []
  const results = []
  const additional = []
  const fcfsSales = []
  listings.forEach((listing) => {
    if (isFcfsSalesListing(listing)) {
      fcfsSales.push(listing)
    } else if (dayjs(listing.Application_Due_Date) > dayjs()) {
      if (!filters || listing.Does_Match) {
        setMatch(true)
        open.push(listing)
      } else {
        additional.push(listing)
      }
    } else {
      if (isLotteryComplete(listing)) {
        results.push(listing)
      } else {
        upcoming.push(listing)
      }
    }
  })
 ```

Then we just pass them into the wrapper function and render right below the open listings
```
 {fcfsSalesView(
                listings.fcfsSales,
                props.directoryType,
                props.getSummaryTable,
                hasFiltersSet,
                isSalesFcfsEnabled
              )}
```

Comment (tallulahkay) [reply app/javascript/modules/listings/DirectoryHelpers.tsx:251]: ah, gotcha!

**Review by tallulahkay — APPROVED** [review]
This looks good! I like how you separated out the two fcfs states into separate lists - good future thinking for the next directory phase. Can you get it PA tested on the review environment (before you merge)

**Comment by tallulahkay** [comment]
Somebody else should review this bc my pr has been merged in so I'm a co-author. I'm removing my review!

File: app/javascript/modules/listings/DirectoryHelpers.tsx
Comment (tallulahkay) [inline app/javascript/modules/listings/DirectoryHelpers.tsx:396]: Remove `isSalesFcfsEnabled` from this if statement

**Review by alulabeshue-sfgov — APPROVED** [review]
LGTM. We've discussed refactoring and adding code coverage in [DAH-2947](https://sfgovdt.jira.com/browse/DAH-2947)

File: app/javascript/modules/listings/DirectoryHelpers.tsx
Comment (cade-exygy) [reply app/javascript/modules/listings/DirectoryHelpers.tsx:373]: Yeah, I will look into it, I think either works, but I think your suggestion is cleaner

**Pull request merged**
